doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1658 @@
+
+    uint64_t (*get_caps)(const OSSL_POLL_GROUP *self);
+    int (*poll)(OSSL_POLL_METHOD *self, /* as shown for SSL_poll */);
+    OSSL_POLL_GROUP *(*create_poll_group)(OSSL_POLL_METHOD *self,
+                                          const OSSL_PARAM *params);
+} OSSL_POLL_METHOD_FUNCS;
+
+OSSL_POLL_METHOD *OSSL_POLL_METHOD_new(const OSSL_POLL_METHOD_FUNCS *funcs,
+                                       size_t funcs_len, size_t data_len);
+
+void *OSSL_POLL_METHOD_get0_data(const OSSL_POLL_METHOD *self);
+
+int OSSL_POLL_METHOD_free(OSSL_POLL_METHOD *self);
+void OSSL_POLL_METHOD_do_free(OSSL_POLL_METHOD *self);
+int OSSL_POLL_METHOD_up_ref(OSSL_POLL_METHOD *self);
+
+uint64_t OSSL_POLL_METHOD_get_caps(const OSSL_POLL_METHOD *self);
+int OSSL_POLL_METHOD_supports_poll_descriptor(OSSL_POLL_METHOD *self,
+                                              const BIO_POLL_DESCRIPTOR *desc);
+int OSSL_POLL_METHOD_poll(OSSL_POLL_METHOD *self, ...);
+OSSL_POLL_GROUP *OSSL_POLL_METHOD_create_poll_group(OSSL_POLL_METHOD *self,
+                                                   const OSSL_PARAM *params);
+
+typedef struct ossl_poll_group_funcs_st {
+    int (*free)(OSSL_POLL_GROUP *self);
+    int (*up_ref)(OSSL_POLL_GROUP *self);
+
+    int (*change_poll)(OSSL_POLL_GROUP *self, /* as shown for change_poll */);
+} OSSL_POLL_GROUP_FUNCS;
+
+OSSL_POLL_GROUP *OSSL_POLL_GROUP_new(const OSSL_POLL_GROUP_FUNCS *funcs,
+                                     size_t funcs_len, size_t data_len);
+void *OSSL_POLL_GROUP_get0_data(const OSSL_POLL_GROUP *self);
+
+int OSSL_POLL_GROUP_free(OSSL_POLL_GROUP *self);
+int OSSL_POLL_GROUP_up_ref(OSSL_POLL_GROUP *self);
+int OSSL_POLL_GROUP_change_poll(OSSL_POLL_GROUP *self,
+                                /* as shown for change_poll */);
+```
+
+Here is how an application might define and create a `OSSL_POLL_METHOD` instance
+of its own:
+
+```c
+struct app_poll_method_st {
+    uint32_t refcount;
+} APP_POLL_METHOD;
+
+static int app_poll_method_free(OSSL_POLL_METHOD *self)
+{
+    APP_POLL_METHOD *data = OSSL_POLL_METHOD_get0_data(self);
+
+    if (!--data->refcount)
+        OSSL_POLL_METHOD_do_free(self);
+
+    return 1;
+}
+
+static int app_poll_method_up_ref(OSSL_POLL_METHOD *self)
+{
+    APP_POLL_METHOD *data = OSSL_POLL_METHOD_get0_data(self);
+
+    ++data->refcount;
+
+    return 1;
+}
+
+static uint64_t app_poll_method_get_caps(const OSSL_POLL_METHOD *self)
+{
+    return OSSL_POLL_METHOD_CAP_IMMEDIATE;
+}
+
+static int app_poll_method_supports_poll_descriptor(OSSL_POLL_METHOD *self,
+                                                    const BIO_POLL_DESCRIPTOR *d)
+{
+    return d->type == BIO_POLL_DESCRIPTOR_TYPE_SOCK_FD;
+}
+
+/* etc. */
+
+OSSL_POLL_METHOD *app_create_custom_poll_method(void)
+{
+    OSSL_POLL_METHOD *self;
+    APP_POLL_METHOD *data;
+
+    static const OSSL_POLL_METHOD_FUNCS funcs = {
+        app_poll_method_free,
+        app_poll_method_up_ref,
+        app_poll_method_get_caps,
```

Comment:
Is this needed, could this be just inferred from whether poll() and/or create_poll_group() functions are set?

Reply:
This was considered but decided against. Using NULL function pointers to determine non-support for something can create issues if someone wants to forward to an underlying implementation, since whether an implementation method's function pointer is NULL or not isn't really something a method can change dynamically. Support might also vary dynamically depending on the runtime environment.

Reply:
Will we consider a NULL pointer for the function that is supposedly implemented according to caps an application error and not NULL-check the pointer before calling then?